File: src/wsclients/r/EPEPT_AddamaWSClient_win.r
print("Begin Execution -> Init Permutation input variables and Load JSON Library");
#init
library(rjson);

args = '';
args[1] = "{ci:'95',oopt_chk:'false',ci_chk:'true',cc_chk:'false',method:'PWM'}"
args[2] = "C:\\jakelin\\perm_R\\out"
args[3] = "http://informatics.systemsbiology.net"
args[4] = "C:\\jakelin\\perm_R\\jl_matrix.tsv"
print(args);

splittedFilePath = strsplit(args[4], "\\", fixed=TRUE);
mylength = length(splittedFilePath[[1]]);
myOutFile = splittedFilePath[[1]][mylength];
splittedFileName = strsplit(myOutFile, ".", fixed=TRUE);
myOutFile = splittedFileName[[1]][1];
myOutFile = paste(myOutFile, ".txt", sep="");

epept_params = args[1]; #"{\"ci\":'95',\"oopt_chk\":'false',\"ci_chk\":'true',\"cc_chk\":'false',\"method\":\"PWM\"}";
host = args[3]; #<- paste("http://cornea.systemsbiology.net:9080");
addama_bot_path <- paste("/addama-rest/primary-repo/path/RobotForms/EPEPT/form_entry");
pOutput = args[2];
file_tsv <- paste("\"filename=@", args[4], "\"", sep="");
myURI = "";
runningStatus = "running";
iter = 1;

epept_makerequest <- function (file_tsv, params, host, addama) {
   curlPostCommand <- paste("curl -X POST -F ", file_tsv, " -F JSON=", epept_params," ", host, addama_bot_path,sep="");
   print(curlPostCommand);
   charVectorPostFormEntry <- try(shell(curlPostCommand, intern=TRUE, ignore.stderr=TRUE));
   print(charVectorPostFormEntry);
   curlPostJson <- fromJSON(charVectorPostFormEntry);
   postedUri <- curlPostJson$uri;
   postUriMsg <- paste('Curl Posted and return this uri:',postedUri);
   return(postedUri);
}

epept_checkstatus <- function (host, uri) {
   curlGetCommand <- paste("curl -X GET ", host, uri, "/status", sep="");
   curlGetStatus <- try(shell(curlGetCommand, intern=TRUE, ignore.stderr=TRUE));
   curlGetStatusJson <- fromJSON(curlGetStatus);
   curlGetStatusChildren <- curlGetStatusJson$children;
   len <- length(curlGetStatusChildren);
   status = 'running';
   for(i in 1:len){
      child <- curlGetStatusChildren[[i]];
      out <- paste(child$name, child$uri, sep=" ");
      if (child$name == 'error'){
          status = 'error';
          break;
      }
      if (child$name == 'completed'){
          status = 'completed';
          break;
      }
      print(out);
   }
   return(status);
}

#Process logs/output
epept_getoutput <- function(host, uri, status){
   nodeType <- paste("/outputs");
   if (status == "error"){
      nodeType <- paste("/logs");
   }
   curlGetOutputCommand <- paste("curl -X GET ", host, uri, nodeType, sep="");
   curlGetStatus <- try(shell(curlGetOutputCommand, intern=TRUE,ignore.stderr=TRUE));
   curlGetStatusJson <- fromJSON(curlGetStatus);
   curlGetStatusChildren <- curlGetStatusJson$children;
   childrenLen <- length(curlGetStatusChildren);
   uri_splitted <- unlist(strsplit(uri, "/"));
   len <- length(uri_splitted);
   form_entry_node <- uri_splitted[len];
   for(i in 1:childrenLen){
      child <- curlGetStatusChildren[[i]];
      #out <- paste(child$name, child$uri, sep=" ");
      curlGetOutputCommand <- paste('curl -X GET ', host, uri, nodeType, '/',child$name, ' >> ', pOutput,  '\\',form_entry_node, '_', child$name, '',  sep='');
      print(curlGetOutputCommand);
      curlGetStatus <- try(shell(curlGetOutputCommand,intern=TRUE,ignore.stderr=TRUE));
   }
   if (status == 'completed'){
      pValueFile <- paste(pOutput, "\\", form_entry_node, "_Pvalues_", myOutFile,sep="");
      outMsg <- paste("PValue matrix has been saved to ", pValueFile,  sep="");
      print(outMsg);
      #try(shell(paste("notepad ", pValueFile), intern=TRUE));       
      #myPVal_Matrix <- matrix(scan("/local/temp/matrix.tsv"), 3, 4, byrow=TRUE)
      pFile <- file(pValueFile);
      cat(readLines(pFile));
     
      #pValueReadIn <- read.table(pValueFile, sep="\t");
      #pMatrix <- as.matrix(pValueReadIn);
      cat("\n\nEPEPT completed successfully, see your output folder for results", "\n", pValueFile);
      return(pValueFile);
   }
   logMsg <- paste("Error, check logs at", oOutput, sep=" "); 
   return(logMsg);
}

#Make request
myURI <- epept_makerequest(file_tsv, perm_params, host, addama_bot_path);

while (runningStatus == 'running') {
   runningStatus <- epept_checkstatus(host, myURI);
   statusMsg <- paste("iteration ", iter, runningStatus, sep=" ");
   if (iter > 5){
      break;
   }
   iter <- iter + 1;
   Sys.sleep(3);
}

EPEPT_Out <- epept_getoutput(host, myURI, runningStatus);

File: src/wsclients/ruby/EPEPT_AddamaWSClient_win.rb
require 'rubygems'
require 'httpclient' # sudo gem install httpclient
require 'json' # sudo gem install json
class EPEPT_AddamaWSClient

#Initialize host and parameters
def initialize()
  puts "Begin initializing parameters\n";
  #@argv_array
  @argv_array = Array.new;
  @argv_array << "C:\\EPEPT\\data\\yeastdata.tsv";
  @argv_array << "http://informatics.systemsbiology.net";
  @argv_array << "C:\\EPEPT\\out";
  @argv_array << "{mode:'PV',method:'PWM',ci_chk:'true',ci:'50',oopt_chk:'false',cc_chk:'false',mail_address:'jlin212@gmail.com'}";
  puts "my arguments " + @argv_array[0] + @argv_array[1] +  @argv_array[2] + @argv_array[3];
  @clnt = HTTPClient.new
  @file_tsv = @argv_array[0]; #'./mymatrix.tsv'
  @HOST = @argv_array[1]; #'http://cornea.systemsbiology.net:9080'
  @outpath = @argv_array[2];
  @params = @argv_array[3]; #"{method:'PWM',ci_chk:'true',ci:'50',oopt_chk:'false',cc_chk:'false'}"  
  @ADDAMA_BOT_PATH = '/addama-rest/primary-repo/path/RobotForms/EPEPT/form_entry';
  @status = 'pending';
  @outNode = 'outputs';
  @formName = "" 
  filePathSplitted = @file_tsv.split("\\");
  @myOutFile = filePathSplitted.last;  
  rescue Exception:
     puts "Initialize Function failed, please check your input arguments!";
end

#Checks status after making request
def epept_checkstatus?(uri)
   getResp = @clnt.get(@HOST + uri + '/status');
   #puts getResp.content
   jsonResp = JSON.parse(getResp.content);
   jsonChild = jsonResp['children']
   jsonChild.each() do |jc|
      cnode_name = jc['name']
      if (cnode_name == 'error') then
         @status = 'error';
         @outNode = 'logs';
         puts 'Errors -> Program exited with errors: see the log files that are being saved to your execution path'
         return false;
      elsif (cnode_name == 'completed') then
         @status = 'completed';
         puts 'Job Completed -> The output files (plot.png and Pvalues_mymatrix.tsv) are saved are saved to your execution path'
         return false;
      end
   end
   return true;
end

#Processes output after job completes
def epept_getoutput(uri, nodeType)
  puts @HOST + uri + '/' + nodeType;
  resp = @clnt.get(@HOST + uri + '/' + nodeType);
  jsonResp = JSON.parse(resp.content);
  jsonChild = jsonResp['children'];
  jsonChild.each() do |child|
   puts 'saving:' + child['name']
   fileResp = @clnt.get(@HOST + uri + '/' + nodeType + '/' + child['name']);
   open(@outpath +"/"+ @formName + "_" + child['name'], "wb") { |file|
    file.write(fileResp.content)
   }
  end
end

#Posts request
def epept_makerequest()
#@clnt.set_cookie_store("cookie.dat")
File.open(@file_tsv) do |file|
#   body = { 'upload' => file, 'filename' => 'mymatrix.tsv', 'method' => 'PWM', 'lower_bound' => '95', 'upper_bound' => '5', 'ci' => 'true', 'oopt' => 'false' }
   body = { 'filename' => file, 'JSON' => @params } 
   res = @clnt.post(@HOST + @ADDAMA_BOT_PATH, body)
   jsonInst = JSON.parse(res.content)
   puts 'uri=>' + jsonInst['uri']
   @formName = jsonInst['name']
   iter = 0
   while (epept_checkstatus?(jsonInst['uri']) && (iter <= 10)) do
      puts 'still running, sleep 3 seconds -> poll at most 10 times'
      sleep 4
      iter = iter + 1;
   end   
      puts 'Done running, job status ->' + @status + '-> processingOutput'
      epept_getoutput(jsonInst['uri'], @outNode)      
  end
 end
 rescue Exception:
      puts "Failed making request, please check if you have a valid internet connection and valid input arguments!";

end

#Invokes program
epeptWS = EPEPT_AddamaWSClient.new
epeptWS.epept_makerequest()
